analytical_engine/apps/simple_path/all_simple_paths.h

```diff
@@ -0,0 +1,324 @@
+/** Copyright 2020 Alibaba Group Holding Limited.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+Author: Ma JingYuan<nn9902@qq.com>
+*/
+
+#ifndef ANALYTICAL_ENGINE_APPS_SIMPLE_PATH_ALL_SIMPLE_PATHS_H_
+#define ANALYTICAL_ENGINE_APPS_SIMPLE_PATH_ALL_SIMPLE_PATHS_H_
+
+#include <set>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "simple_path/all_simple_paths_context.h"
+
+#include "core/app/app_base.h"
+#include "core/worker/default_worker.h"
+
+namespace gs {
+
+template <typename FRAG_T>
+class AllSimplePaths : public AppBase<FRAG_T, AllSimplePathsContext<FRAG_T>>,
+                       public grape::Communicator {
+ public:
+  INSTALL_DEFAULT_WORKER(AllSimplePaths<FRAG_T>, AllSimplePathsContext<FRAG_T>,
+                         FRAG_T)
+  static constexpr grape::MessageStrategy message_strategy =
+      grape::MessageStrategy::kAlongOutgoingEdgeToOuterVertex;
+  static constexpr grape::LoadStrategy load_strategy =
+      grape::LoadStrategy::kBothOutIn;
+  using vertex_t = typename fragment_t::vertex_t;
+  using vid_t = typename fragment_t::vid_t;
+
+  void PEval(const fragment_t& frag, context_t& ctx,
+             message_manager_t& messages) {
+    vertex_t source;
+    vid_t source_gid;
+    bool empty_flag = false;
+    // Source does not exist.
+    if (!frag.Oid2Gid(ctx.source_id, source_gid)) {
+      empty_flag = true;
+      ctx.nodes_not_found = true;
+    }
+    // nodes_not_found = true, return 2 as NodeNotFound.
+    if (ctx.nodes_not_found == true && frag.fid() == 0) {
+      std::vector<typename fragment_t::oid_t> data;
+      std::vector<size_t> shape{1};
+      data.push_back(2);
+      ctx.assign(data, shape);
+      return;
+    }
+    ctx.soucre_fid = source_gid >> ctx.fid_offset;
+    if (ctx.targets.count(source_gid) || ctx.cutoff < 1) {
+      empty_flag = true;
+    }
+    // empty_flag == true, return 1 as empty path.
+    if (empty_flag == true && frag.fid() == ctx.soucre_fid) {
+      std::vector<typename fragment_t::oid_t> data;
+      std::vector<size_t> shape{1};
+      data.push_back(1);
+      ctx.assign(data, shape);
+      return;
+    } else if (empty_flag == true || ctx.nodes_not_found == true) {
+      return;
+    }
+    bool native_source = frag.GetInnerVertex(ctx.source_id, source);
+    // If native_source resize edge_map, else send frag innervertices num to
```

Comment:
If is native_source, resize the edge_map, otherwise send the fragment inner vertex num to source-vertex's fragment